src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorProjectChangePublisher.cs

```diff
@@ -341,17 +349,32 @@ protected virtual bool ShouldSerialize(ProjectSnapshot projectSnapshot, string c
             return !status.IsInProgress && _documentsProcessed;
         }
 
+        private void ImmediatePublish(ProjectSnapshot projectSnapshot)
+        {
+            lock (_pendingProjectPublishesLock)
+            {
+                // Clear any pending publish
+                _pendingProjectPublishes.Remove(projectSnapshot.FilePath);
```

Comment:
This is the actual fix. Gist of what was happening before was when you removed a document two events would fire: `DocumentRemoved`, `ProjectChanged`. 

First would be `DocumentRemoved` which wouldn't know about the new TagHelper state (it'd be in the process of being calculated), it uses delay publishing because in old-world, pre-SolutionClosing awareness, it'd slam the project system on close. I wonder if we can change this now, anyhow that's the state of the world as-of-today and there's lots of other pending publishes that can happen.

Next would be `ProjectChanged` which would be triggered off of the change of TagHelpers. `ProjectChanged` would be triggered immediately updating project state and publishing a new project.razor.json with the -1 TagHelper included.

After a delay the `DocumentRemoved` trigger would fire and would then re-publish a project.razor.json with the old TagHelper state causing the stale TagHelper.